source/new-image-handler/src/processor/image/watermark.ts

```diff
@@ -0,0 +1,288 @@
+        width = metadata.width - 1;
+        needResize = true;
+      }
+
+      if (markMetadata.height !== undefined && metadata.height !== undefined && markMetadata.height > metadata.height) {
+        height = metadata.height - 1;
+        needResize = true;
+      }
+      if (needResize) {
+        watermarkImg = watermarkImg.resize(width,height);
+      }
+    }
+    const bt = await watermarkImg.toBuffer();
+    ctx.image.composite([{ input: bt, tile: opt.fill, gravity: opt.g }]);
+  }
+
+  gravityConvert(param: string): string {
+    if (['north', 'west', 'east', 'south', 'center', 'centre', 'southeast', 'southwest', 'northwest'].includes(param)) {
+      return param;
+    } else if (param === 'se') {
+      return 'southeast';
+    } else if (param === 'sw') {
+      return 'southwest';
+    } else if (param === 'nw') {
+      return 'northwest';
+    } else if (param === 'ne') {
+      return 'northeast';
+    } else {
+      throw new InvalidArgument('Watermark param \'g\' must be in \'north\', \'west\', \'east\', \'south\', \'center\', \'centre\', \'southeast\', \'southwest\', \'northwest\'');
+    }
+  }
+
+  calculateTextSize(text: string, fontSize: number): WatermarkTextOpts {
+    let cWidth = 0;
+    for (let v of text) {
+      const charCode = v.charCodeAt(0);
+      if (charCode > 256) {
+        cWidth += fontSize;
+      } else if (charCode > 97) {
+        cWidth += fontSize / 2;
+      } else {
+        cWidth += fontSize * 0.8;
+      }
+    }
+    return {
+      width: Math.round(cWidth + margin),
+      height: Math.round(fontSize * 1.4),
+    };
+  }
+  textSvgStr(opt: WatermarkOpts, textOpt: WatermarkTextOpts): string {
+    const xOffset = Math.round(textOpt.width / 2);
+    const yOffset = Math.round(textOpt.height * 0.8);
+    const color = `#${opt.color}`;
+    const opacity = opt.t / 100;
+    const svg = `<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 ${textOpt.width} ${textOpt.height}" text-anchor="middle">
+    <text font-size='${opt.size}'  x="${xOffset}" y="${yOffset}" fill="${color}" opacity="${opacity}">${opt.text}</text>
+    </svg>`;
+    return svg;
+  }
+
+  textSvgImg(svgBytes: Buffer, textOpt: WatermarkTextOpts): sharp.Sharp {
+    const overlapImg = sharp({
+      create: {
+        width: textOpt.width + margin,
+        height: textOpt.height + margin,
+        channels: 4,
+        background: { r: 0, g: 0, b: 0, alpha: 0 },
+      },
+    }).composite([{ input: svgBytes }]);
+
+    return overlapImg;
+  }
+
+  async autoResizeImg(source: sharp.Sharp, ctx: IImageContext, opt: WatermarkOpts, textOpt: WatermarkTextOpts): Promise<sharp.Sharp> {
+    if (opt.auto) {
+
+      let w = textOpt.width;
+      let h = textOpt.height;
+      let needResize = false;
+      const overlapImgMeta = await source.metadata();
+      const metadata = await ctx.image.metadata();
+
+      if (overlapImgMeta.width !== undefined && metadata.width !== undefined && overlapImgMeta.width > metadata.width) {
```

Comment:
Do this`if (overlapImgMeta.width && metadata.width && overlapImgMeta.width > metadata.width)`